core/modules/memman/MemManReal.cc

```diff
@@ -113,7 +117,55 @@ MemMan::Status MemManReal::getStatus(Handle handle) {
 /*                                  l o c k                                   */
 /******************************************************************************/
   
-MemMan::Handle MemManReal::lock(std::vector<TableInfo> const& tables, int chunk) {
+int MemManReal::lock(MemMan::Handle handle, bool strict) {
+
+    MemFileSet* fsP = nullptr;
+    int rc;
+
+    // If this is a nil handle, then we need not do anything more. If this is
+    // a bad handle, return failure.
+    //
+    if (handle == HandleType::ISEMPTY) return 0;
+    if (handle == HandleType::INVALID) return EINVAL;
+
+    // Find the table set in the set cache. We need the handle mutex to do this
+    // so we use a lock gaurd in a nested scope. Since we don't want
+    // to keep the handle mutex during the long running lock operation, we
+    // get the pointer to he file set and lock it prior to leaving the scope.
+    // If the file set is already locked, then a lock() call is in progress.
+    //
+    {    std::lock_guard<std::mutex> guard(hanMutex);
+         auto it = hanCache.find(handle);
+         if (it == hanCache.end() || !(it->second->isOwner(_memory))) {
+             return ENOENT;
+         }
+         fsP = it->second;
+         fsP->serialize(true);
+         _numLocks++;
+    }
+
+    // Perform the lock and then drop the file set lock.
+    //
+    rc = fsP->lockAll(strict);
+    fsP->serialize(false);
```

Comment:
Well, simplicity may be in the eye of the beholder, That particular lock 
is used to gaurd against multiple operations on the same file set that 
also happens to includes delete. Yes, I could go around replacing simple 
locks everywhere with shared pointers and additional gaurd locks. I 
actually find that more complicated and not very obvious on what the 
control flow is. Also, it would require substantial lock reorganization to 
obtain the same amount of parallelsm I wanted to maintain.

On Mon, 11 Jul 2016, John Gates wrote:

>> +    // If the file set is already locked, then a lock() call is in progress.
>> +    //
>> +    {    std::lock_guard<std::mutex> guard(hanMutex);
>> +         auto it = hanCache.find(handle);
>> +         if (it == hanCache.end() || !(it->second->isOwner(_memory))) {
>> +             return ENOENT;
>> +         }
>> +         fsP = it->second;
>> +         fsP->serialize(true);
>> +         _numLocks++;
>> +    }
>> +
>> +    // Perform the lock and then drop the file set lock.
>> +    //
>> +    rc = fsP->lockAll(strict);
>> +    fsP->serialize(false);
>
> fsP->serialize() seems overly complicated.
>
> If fsp was a shared_ptr and hanCache used shared_ptr, you wouldn't need to call delete in MemManReal::unlock(), and wouldn't need to call serialize(false) in MemFileSet::~MemFileSet() as it wouldn't be deleted until everyone was done with it.
>
> The code here could be changed to the following and MemFileSet::serialize(bool) can go away.
> ```
>     {
>        std::unique_lock<std::mutex> hanGuard(hanMutex);
>        ///  ... find handle in hanCache ...
>        fsP = it->second;
>        std::lock_guard<std::mutex> fspGuard(fsP->getMutex());  // return a reference to _setmutex
>        _numlocks++;
>        hanGuard.unlock();
>        rc = fsP->lockAll(strict);
>     }
> ```
> The main advantage here being there's no way you can get out of that block without unlocking the mutexes, nor can the MemFileSet destructor accidentally be called when some other thread has locked _setMutex.
>
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/lsst/qserv/pull/238/files/b8110b88ccb31da5b6475bda78100475fadbb710#r70283114